File: Assets/Bundles/Client/Resources/TS/Controllers/Global/Team/TeamController.lua
-- Compiled with unity-ts v2.1.0-75
local Reflect = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Reflect
local Controller = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Controller
local Object = require("Shared/rbxts_include/node_modules/@easy-games/unity-object-utils/init")
local Network = require("Shared/TS/Network").Network
local Team = require("Shared/TS/Team/Team").Team
local TeamController
do
	TeamController = setmetatable({}, {
		__tostring = function()
			return "TeamController"
		end,
	})
	TeamController.__index = TeamController
	function TeamController.new(...)
		local self = setmetatable({}, TeamController)
		return self:constructor(...) or self
	end
	function TeamController:constructor(playerController)
		self.playerController = playerController
		self.teams = {}
	end
	function TeamController:OnStart()
		Network.ServerToClient.AddTeams.Client:OnServerEvent(function(teamDtos)
			for _, dto in teamDtos do
				local team = Team.new(dto.name, dto.id, Color.new(dto.color[1], dto.color[2], dto.color[3], dto.color[4]))
				local _teams = self.teams
				local _id = dto.id
				_teams[_id] = team
				for _1, userId in dto.userIds do
					local player = self.playerController:GetPlayerFromUserId(userId)
					if player then
						team:AddPlayer(player)
					end
				end
			end
		end)
		Network.ServerToClient.RemoveTeams.Client:OnServerEvent(function(teamIds)
			for _, teamId in teamIds do
				local team = self:GetTeam(teamId)
				if not team then
					continue
				end
				local _teams = self.teams
				local _teamId = teamId
				_teams[_teamId] = nil
			end
		end)
		Network.ServerToClient.AddPlayerToTeam.Client:OnServerEvent(function(teamId, userId)
			local team = self:GetTeam(teamId)
			if not team then
				return nil
			end
			local player = self.playerController:GetPlayerFromUserId(userId)
			if not player then
				return nil
			end
			team:AddPlayer(player)
		end)
		Network.ServerToClient.RemovePlayerFromTeam.Client:OnServerEvent(function(teamId, playerId)
			local team = self:GetTeam(teamId)
			if not team then
				return nil
			end
			local player = self.playerController:GetPlayerFromUserId(playerId)
			if not player then
				return nil
			end
			team:RemovePlayer(player)
		end)
	end
	function TeamController:GetTeam(teamId)
		local _teams = self.teams
		local _teamId = teamId
		return _teams[_teamId]
	end
	function TeamController:GetTeams()
		return Object.values(self.teams)
	end
end
-- (Flamework) TeamController metadata
Reflect.defineMetadata(TeamController, "identifier", "Bundles/Client/Controllers/Global/Team/TeamController@TeamController")
Reflect.defineMetadata(TeamController, "flamework:parameters", { "Bundles/Client/Controllers/Global/Player/PlayerController@PlayerController" })
Reflect.defineMetadata(TeamController, "flamework:implements", { "$:flamework@OnStart" })
Reflect.decorate(TeamController, "$:flamework@Controller", Controller, { {} })
return {
	TeamController = TeamController,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Bundles/Imports/Core/Core~/src/Client/Controllers/Team/TeamController.ts
import { Controller, OnStart } from "@easy-games/flamework-core";
import Object from "@easy-games/unity-object-utils";
import { CoreNetwork } from "Shared/CoreNetwork";
import { Team } from "Shared/Team/Team";
import { PlayerController } from "../Player/PlayerController";

@Controller({})
export class TeamController implements OnStart {
	private teams = new Map<string, Team>();

	constructor(private readonly playerController: PlayerController) {}

	OnStart(): void {
		CoreNetwork.ServerToClient.AddTeams.Client.OnServerEvent((teamDtos) => {
			for (let dto of teamDtos) {
				const team = new Team(
					dto.name,
					dto.id,
					new Color(dto.color[0], dto.color[1], dto.color[2], dto.color[3]),
				);
				this.teams.set(dto.id, team);
				for (let userId of dto.userIds) {
					const player = this.playerController.GetPlayerFromUserId(userId);
					if (player) {
						team.AddPlayer(player);
					}
				}
			}
		});

		CoreNetwork.ServerToClient.RemoveTeams.Client.OnServerEvent((teamIds) => {
			for (let teamId of teamIds) {
				const team = this.GetTeam(teamId);
				if (!team) continue;

				this.teams.delete(teamId);
			}
		});

		CoreNetwork.ServerToClient.AddPlayerToTeam.Client.OnServerEvent((teamId, userId) => {
			const team = this.GetTeam(teamId);
			if (!team) return;

			const player = this.playerController.GetPlayerFromUserId(userId);
			if (!player) return;

			team.AddPlayer(player);
		});

		CoreNetwork.ServerToClient.RemovePlayerFromTeam.Client.OnServerEvent((teamId, playerId) => {
			const team = this.GetTeam(teamId);
			if (!team) return;

			const player = this.playerController.GetPlayerFromUserId(playerId);
			if (!player) return;

			team.RemovePlayer(player);
		});
	}

	public GetTeam(teamId: string): Team | undefined {
		return this.teams.get(teamId);
	}

	public GetTeams(): Team[] {
		return Object.values(this.teams);
	}
}
